docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+commit = proc.stdout.strip()
+link = f"https://github.com/ibis-project/ibis/tree/{commit}"
+
+versions = pd.DataFrame(
+    [("Python", sys.version)] + sorted(
+        [
+            *map(version_pair, (pd, dask, distributed, pa, duckdb)),
+            ("ibis", f"[`{commit}`]({link})"),
+        ]
+    ),
+    columns=["Dependency", "Version"],
+)
+
+print(versions.to_markdown(index=False))
+```
+
+### Data
+
+I used the files in this link to run my experiment: https://raw.githubusercontent.com/pypi-data/data/20135ed214be9d6bb9c316121e5ccdaf29c6b9b1/links/dataset.txt
+
+Here's a summary of the data set's file sizes:
+
+```{python}
+#| echo: false
+#| output: asis
+import glob
+import subprocess
+
+allfiles = glob.glob("/data/pypi-parquet/*.parquet")
+cmd = ("du", "-h", *allfiles)
+
+proc = subprocess.run(cmd, check=True, capture_output=True, text=True)
+print(f"""```bash\n$ du -h /data/pypi-parquet/*.parquet\n```""")
+print(f"""```\n{proc.stdout}\n```""")
+```
+
+## Recapping the original Ibis post
+
+Check out [the original blog
+post](https://ibis-project.org/posts/querying-pypi-metadata-compiled-languages/)
+if you haven't already!
+
+Here's the Ibis + DuckDB code, along with a timed execution of the query:
+
+```{python}
+#| echo: false
+#| output: asis
+with open("./step0.py") as f:
+    source = f.read()
+
+exec(source)
+print(f"```python\n{source}\n```")
+```
+
+```{python}
+%time df = expr.to_pandas()
+df
+```
+
+1. We've since implemented [a `flatten` method](../../reference/expression-collections.qmd#ibis.expr.types.arrays.ArrayValue.flatten)
+   on array expressions so it's no longer necessary to define a UDF here. I'll
+   leave this code unchanged for this post. **This has no effect on the
+   performance of the query**. In both cases the generated code contains
+   a DuckDB-native call to [its `flatten`
+   function](https://duckdb.org/docs/sql/functions/nested.html).
+2. This is a small change from the original query that adds a final sort key to
+   make the results deterministic.
+
+## Pandas
+
+Let's try to replicate this workflow using pandas.
+
+I started with this code:
+
+```{python}
+#| error: true
+import pandas as pd
+
+df = pd.read_parquet("/data/pypi-parquet/*.parquet")
+```
+
+Looks like pandas doesn't support globs. That's fine, we can use the builtin
+`glob` module.
+
+```python
+import glob
+
+df = pd.read_parquet(glob.glob("/data/pypi-parquet/*.parquet"))
+```
+
+This eventually triggers the [Linux OOM
+killer](https://lwn.net/Kernel/Index/#Memory_management-Out-of-memory_handling)
+after some minutes, so I can't run the code.
+
+Let's try again with just a single file. I'll pick the smallest file, to avoid any
+potential issues with memory and give pandas the best possible shot.
+
+The [smallest file](#data-size) is 632 MiB on disk.
+
+```{python}
+import os
+
+smallest_file = min(glob.glob("/data/pypi-parquet/*.parquet"), key=os.path.getsize)
+
+%time df = pd.read_parquet(smallest_file)
```

Comment:
please use ``dtype_backend="pyarrow"`` or set ``pd.options.infer_string = True``, the latter will be the default in pandas 3.0

Converting from arrow strings to objects takes up approx 80% of your runtime if you have string columns (see https://medium.com/towards-data-science/utilizing-pyarrow-to-improve-pandas-and-dask-workflows-2891d3d96d2b in dtype_backend in PyArrow-native I/O readers)

This will improve all your string operations by a huge factor as well

Reply:
Sweet, giving this a try.

Thanks for the tip!

Reply:
Indeed this is much faster for loading. I'll report back if there are any issues running the query, or push the results if all goes well!

Reply:
Just to close this out (I've noted elsewhere too) I wasn't able to use `dtype_backend="pyarrow"` due to:

1. The offset issue
2. Missing implementation of `extract` for pyarrow strings
3. The month snapping workaround not working with pyarrow backend series'